draft-ietf-dnsop-svcb-https.md

```diff
@@ -659,6 +659,31 @@ each RRSet in the Additional section with the same DNSSEC-related records
 that they would send when providing that RRSet as an Answer (e.g. RRSIG, NSEC,
 NSEC3).
 
+## EDNS Client Subnet (ECS)
+
+The EDNS Client Subnet extension (ECS, {{!RFC7871}}) allows recursive
+resolvers to request IP addresses that are suitable for a particular client
+IP range.  IP addresses appear in two places in Authoritative SVCB responses:
+
+* ipv*hint SvcParams in a ServiceMode response.  Clients only use these hints
+when the recursive resolver does not provide A/AAAA records for TargetName in
+the Additional section, as recommended in {{recursive-behavior}}.
+* A/AAAA records returned by the Authoritative server in the Additional
+section of a SVCB response when TargetName is in-zone, as recommended in
+{{authoritative-behavior}}.
+
+A recursive resolver that uses ECS for A/AAAA queries SHOULD send the
+same ECS prefix for SVCB queries unless it knows that any ipv*hint SvcParams
+will not be used, due to the resolver's implementation of Additional section
```

Comment:
I'm not sure I follow the "unless it knows they will not be used".  How does the resolver know that ipv*hint might not be used?

Reply:
Clients only use ipv*hint if the response doesn't include A+AAAA records for TargetName in the Additional section.  If the recursive resolver knows that it is going to include those additional address records in the response, then it knows the hints will never be used.

Reply:
Where this gets confusing is with the subsequent guidance that records in the Additional section should not be included if ECS is being used.   Isn't the guidance effectively then "Send ECS so SVCB records can be usable with ipv*hint.  Plan to not send Additional records." ?

Reply:
I think the proposal is saying: 
when a recursive resolver does not include additional records for the client, it knows hints may be used by the client, so it needs to ensure that the ipv4hint answer is answered properly honoring ECS info. So the recursive needs to send ECS to the authorities. Whether the auths include additionals or not in this case (that is the the subject of "According to Section 7.3.1 of {{!RFC7871}}, " paragraph) does not matter because the recursive is not including the additionals in its response to the client anyway. 

When recursive resolver does include additional records for the client, ipv[46]hint should not be used, so the ipv[46]hint from auths don't need to vary by ECS, hence the /0 suggestion to avoid extra per-subnet caching. Additionally, in this case, the recommendation in the draft is for auths to not return additionals if they think the additionals should also vary by ECS. The recursive, not finding additionals, will subsequently ask A/AAAA queries with proper ECS and will include ECS based answers in in its additional section that it sends to clients. I think the following is implicit: if auths do respond with additionals then recrusives will use those non-ECS based additionals in its response to the client. 

This seems ok but I will note that Section 7.3.1 and 7.4 of RFC7871 does not say auths should not return additionals when receiving an ECS request, or that recursives should not be using the returned additionals from auths, it just prohibits associating the additionals with ECS data. So the recommendation about auths not responding with additionals seems like a new requirement for SVCB handling, it does not follow directly from RFC7871. Was that desired? I am worried that many auths will return non-ECS additionals when sent /0 and it will take time for people to realize and fix this for SVCB.

Reply:
I agree with Mashooq, I don't like the idea of the recursive potentially changing its ECS behavior for SVCB.  That's going to be a mess for debugging, and will potentially trigger some authoritative behavior that tries to infer whether the resolver supports ECS or not.  Sending a source prefix of 0 means "I understand ECS, but I'm not going to use it", and is used as a signal for a desire for privacy on the part of the end user.   I don't like the idea of overloading that for a SVCB-specific meaning.  

RFC7871 is pretty clear about the circumstances under which a resolver that supports ECS would not send it:  a) when it's asking for a TLD/PSL; b) when it has learned that a given auth doesn't support ECS; c) when it is using an explicit ACL and the auth isn't on that list.  (b) is already pretty ambiguous and highly subjective, I'm not excited about adding another exception to that list.

Reply:
@mmuhaime's interpretation matches my intent here.

I don't think the "/0" behavior is too tricky for Authoritative servers: "see ECS -> no Additional" seems simpler than "see ECS with nonzero prefix -> no Additional".  I also don't think that this will take time to sort out.  Neither of these codepaths exists today, so if we clearly specify which behavior nameservers should implement, I think we can get it.  If we don't specify, we might get a mix of both.

@jdreed Having different behavior for SVCB is entirely in line with RFC 7871, which says

>   Implementations MAY also allow additional configuring of this based
>   on other criteria, such as zone or query type.

Sending "/0" means "I support ECS but I have disabled it for this query".  That's exactly how this PR uses it.

Here are some alternative designs:
* ECS recursives send the same ECS for SVCB as they do for A/AAAA.  This significantly reduces cache efficiency when IP hints are included and the recursive resolver knows they will not be used ("full service resolver").
  - Also bind A/AAAA in the Additional section to the ECS scope of the SVCB response.  This reduces latency (no followup query to get ECS A/AAAA) but violates (updates?) the letter of RFC 7871's prohibition against ECS in the Additional section.  (I don't think it violates the spirit: that requirement is for records that will be used by the recursive resolver itself, which these are not.)
     * And also let full service resolvers treat all SVCB records as cacheable for "/0" (i.e. ECS applies _only_ to the Additional section).  This maximizes performance but substantially deviates from RFC 7871.
* Let full service resolvers disable ECS for SVCB, but don't require nameservers to distinguish "/0" from "no ECS".  This has good caching efficiency but adds latency: it requires the resolver to ignore all Additional A/AAAA records if ECS could be used, even when ECS is not actually in use.

Reply:
OK, fair, I think I misread this initially and thought it was proposing one case under which no ECS option would be sent (vs sending a /0), and that was the part I was most worried about.

Reply:
Possible improvement to help with any confusion over "unless it knows they will not be used, due to resolver's implementation of Additional section processing": maybe convert to something along the lines of "unless it implements Additional section processing, and therefore knows they will be unnecessary to a client that properly prefers A/AAAA results over ipv*hint SvcParams".  As written, the statement is kinda backwards in that it explains the justification ("knows they will not be used") before the clear, concise condition ("implements Additional section processing").

Reply:
The original technical proposal seems good to me. I do like @ericorth 's wording better because I think it will make it easier to understand. 
We could also use something like: " If a recursive resolver knows that the client may use the ipv[46]hints, it should use the same ECS prefix for SVCB query that it uses for A/AAAA queries when questioning Authoritative servers.  
 If a recursive resolver knows that the client will not use ipv[46]hints, then it MAY include an ECS with a SOURCE PREFIX-LENGTH of zero to indicate that ECS is supported but disabled. One scenario where the recursive knows ipv[46]hints won't be used is when it includes resolved addresses for the SVCB target names in the additionals section in its response to clients."

Reply:
OK, I've reworked the text for clarity based on the feedback in this thread.